Introduce Form Templates Page

shervElmi opened the pull request:
This PR introduces a new "Form Templates" page within the Formidable Admin Panel. The Form Templates page allows users to either start from scratch with a blank form or choose from a selection of pre-built templates.

## QA URL:
https://qa.formidableforms.com/sherv/wp-admin/admin.php?page=formidable-form-templates

## Testing Instructions:
1. Make sure you have Formidable PRO installed and activated on the **[`introduce-form-templates` branch](https://github.com/Strategy11/formidable-pro/tree/introduce-form-templates)**.
    - **Formidable PRO PR**: [https://github.com/Strategy11/formidable-pro/pull/4483](https://github.com/Strategy11/formidable-pro/pull/4483)
2. Confirm that the designs on the Form Templates page match those in the Figma file. **[Figma Design File](https://www.figma.com/file/x8Epd7CJLLY6JIyPoCMPvc/Create-Form?type=design&node-id=1864-17378&mode=dev)**
4. Review the video introductions (Part One and Part Two) and verify that every part of the feature works as shown.

## Video Introductions:
### Updated Custom Templates and Ready-Made Solutions
https://www.loom.com/share/4657fb8d47c2437d879e807ee851bacb?sid=93082411-a37c-4d31-a762-6cfc400418cb

### Part One
https://www.loom.com/share/d8ed343652dd4ed3b817a8d2d54ae284

### Part Two
https://www.loom.com/share/7fa9d217b98e448b8e1419531f86b710

## Checklist
- [x] Write PHPUnit Tests.
- [x] Test real licenses, including (Basic, Pro, Business, Elite, and expired)

Crabcyborg commented:
@shervElmi It looks like you have a section for QA URL but it's blank.

Could you set up a QA URL so we can have others test this as well? I'd love to get Razvan to test it out.

Thank you!

Crabcyborg commented on classes/controllers/FrmAppController.php:
```diff
@@ -375,61 +376,15 @@ public static function upgrade_overlay_html() {
+	 *
 	 * @return void
 	 */
 	private static function new_form_overlay_html() {
```
@shervElmi It's nice not to leave deprecated functions in the middle of the file.

Rather, moving them down to the bottom where they can be ignored more easily.

Crabcyborg commented on classes/controllers/FrmFormTemplatesController.php:
```diff
@@ -0,0 +1,788 @@
+ */
+
+/**
+ * Copyright (C) 2023 Formidable Forms
```
@shervElmi I don't think we need another license in this file. This seems unnecessary.

shervElmi commented:
@Crabcyborg, sure thing! I've just assigned it to you so you can review my code performance for this Quarterly Review.
I plan to add some commits tomorrow and then provide the QA URL along with other relevant details.

Thank you!

Crabcyborg commented on classes/controllers/FrmFormTemplatesController.php:
```diff
@@ -0,0 +1,788 @@
+		$published_forms = FrmForm::get_published_forms();
+
+		// Get IDs from the published forms for matching.
+		$published_form_ids = array_map(
```
With a basic map like this you can save some code and use `wp_list_pluck` https://developer.wordpress.org/reference/functions/wp_list_pluck/.

`$published_form_ids = wp_list_pluck( $published_forms, 'id' );`

Crabcyborg commented on classes/controllers/FrmFormTemplatesController.php:
```diff
@@ -0,0 +1,788 @@
+			);
+
+			// Add the 'is_favorite' key
+			$template['is_favorite'] = in_array( $template['id'], self::$favorite_templates['custom'] );
```
Is it possible to use `$strict = true` here? Are the IDs always integers?

shervElmi replied:
We can't use `$strict = true` here, as the types of the IDs aren't consistently the same, leading to potential issues with strict type checking.

Crabcyborg commented on classes/controllers/FrmFormTemplatesController.php:
```diff
@@ -0,0 +1,788 @@
+	 *
+	 * @return void
+	 */
+	public static function dequeue_scripts() {
```
@shervElmi Since the same code (minus the page check) is in `FrmApplicationsController.php`, could we share some code between those functions?

shervElmi replied:
Personally, I'm cautious about consolidating this dequeuing logic, as it could make the code harder to extend in the future and may violate object-oriented principles. However, since you have deep knowledge of the plugins and their behaviors, if you identify any specific scripts that should be dequeued across different parts, please provide a list and I can create a dedicated method to handle them.

Crabcyborg commented on classes/controllers/FrmFormsController.php:
```diff
@@ -952,134 +929,48 @@ public static function save_per_page( $save, $option, $value ) {
-		$categories = array_diff( $categories, $exclude_cats );
-		sort( $categories );
-		return $categories;
+		_deprecated_function( __METHOD__, 'x.x' );
```
@shervElmi Since this function is private, there's no need to deprecate it. We can just remove it right away.

Crabcyborg commented on classes/helpers/FrmAppHelper.php:
```diff
@@ -318,6 +318,18 @@ public static function is_preview_page() {
+	 */
+	public static function is_form_templates_page() {
+		$page = self::simple_get( 'page', 'sanitize_title' );
+		return strpos( $page, FrmFormTemplatesController::PAGE_SLUG ) !== false;
```
Would the page be anything but an exact match? Can we just check `$page === FrmFormTemplatesController::PAGE_SLUG` instead of doing a `strpos` check?

Crabcyborg replied:
Also, can we just use `FrmAppHelper::is_admin_page( FrmFormTemplatesController::PAGE_SLUG );` instead anyway? That would be more consistent with how we check in other places.

Crabcyborg commented on classes/helpers/FrmAppHelper.php:
```diff
@@ -1218,6 +1230,52 @@ public static function print_admin_banner( $should_show_lite_upgrade ) {
+	 *     @type string $description The description of the section.
+	 *     @type string $link_text   The text for the link.
+	 *     @type string $link_url    The URL for the link.
+	 *     @type string $user        The required user role to view the section. Default 'administrator'.
```
Why `$user`? It seems to be like this should be `$role`. `$user` should be used only for `WP_User` objects.

Crabcyborg commented on classes/helpers/FrmFormTemplatesHelper.php:
```diff
@@ -0,0 +1,198 @@
+	 * @return string[] Array of categories to ignore.
+	 */
+	public static function ignore_template_categories() {
+		return array( 'Business', 'Elite', 'Personal', 'Creator', 'Basic', 'free' );
```
Can we just re-use `FrmFormsHelper::ignore_template_categories` instead of having a duplicate function.

Crabcyborg replied:
I deleted my previous comment. I didn't realize the other function is deprecated.

I don't know how necessary it is to move that function though. Having the extra deprecated code just makes this messier than if we kept it where it was.

Crabcyborg commented on classes/views/form-templates/index.php:
```diff
@@ -0,0 +1,96 @@
+ */
+
+/**
+ * Copyright (C) 2023 Formidable Forms
```
We don't need to copyright every single view file. This just adds a ton of extra text that isn't necessary.

shervElmi replied:
@Crabcyborg, many big companies like Google, which create WordPress plugins, include these texts in every file for professionalism. Since we are growing, I thought it might make our code look more professional.

By the way, I'll go ahead and remove them. :)

<img width="962" alt="image" src="https://github.com/Strategy11/formidable-forms/assets/69119241/afe0be5f-585e-477c-9bae-251cdc0e8127">

Crabcyborg commented on classes/views/form-templates/index.php:
```diff
@@ -0,0 +1,96 @@
+				// Templates list.
+				require $view_path . 'templates.php';
+				?>
+			</div><!-- #post-body-content -->
```
All of these HTML comments actually get loaded in the HTML. It makes the page data larger than necessary.

It's better to stick to PHP comments.

Crabcyborg reviewed (commented):
From the SVG data it doesn't look like the lock icon SVG was optimized.

<img width="970" alt="Screen Shot 2023-09-28 at 6 42 46 PM" src="https://github.com/Strategy11/formidable-forms/assets/9134515/c4db739a-b2d6-4c58-abfc-267fdbf800b2">

Could you follow our steps here from our Developer Guide on optimizing SVGs? I didn't check the other SVGs, but make sure to optimize all of them.

> SVGs
> The Formidable Lite plugin includes an icons.svg file where we include most of the back-end icons. This file is intended for SVGs that will be used multiple times in the admin area. This is not for svgs that are only used once, and should not be loaded on the front-end.
> 
> When adding any SVGs, run them through minification to cut down on the file size. https://jakearchibald.github.io/svgomg/ works well for this. If this is a <symbol>, replace "symbol" with "svg" before copying and pasting it into this site. Most of the time, it works well to use:
> number precision: 1
> transform precision: 1
> most toggles on except: remove xmlns, remove viewbox, remove style elements
> If the result doesn't look good, the precision may need to be increased.

Crabcyborg commented on js/formidable_admin.js:
```diff
@@ -9982,6 +9465,25 @@ function frmAdminBuildJS() {
 			jQuery( '.frm_submit_no_ajax' ).on( 'click', submitNoAjax );
 
+			// Attach click event listeners to 'Name your form' modal
+			const renameFormButton = document.querySelector( '#frm-rename-form-button' );
```
It's nice to use `getElementById` when querying by ID. It's a bit more clear/readable, and it's more optimized.

Crabcyborg commented on js/form-templates/src/utils/validation.js:
```diff
@@ -0,0 +1,82 @@
+ * @return {boolean} True if the template is locked, otherwise false.
+ */
+export const isLockedTemplate = template =>
+	template?.classList.contains( `${ PREFIX }-locked-item` );
```
It isn't great to make everything nullable everywhere. You end up getting weird results.

For example, if `template` is null, `isLockedTemplate( template )` returns `undefined`.

Crabcyborg commented on js/form-templates/src/ui/showModal.js:
```diff
@@ -0,0 +1,139 @@
+
+// Mapping each plan to the subsequent plans it can upgrade to
+const upgradablePlans = {
+	'basic': [ 'basic', 'plus', 'business', 'elite' ],
```
Let's remove the apostrophes on these keys. There isn't usually a reason to wrap the quotes around keys.

Crabcyborg commented on js/form-templates/src/ui/initializeModal.js:
```diff
@@ -0,0 +1,104 @@
+		formHtml = formHtml.replace( regex, '' );
+
+		// Inject form HTML into the email form container
+		leaveEmailModalApiEmailForm.innerHTML = formHtml;
```
@shervElmi Can we put less code in the `try/catch`? It's nice not to put large blocks in a `try/catch`. Just the things we expect to throw an exception.

Crabcyborg commented on js/form-templates/src/ui/initializeModal.js:
```diff
@@ -0,0 +1,104 @@
+
+	try {
+		// Fetch form HTML
+		const response = await fetch( url );
```
Can we use `frmDom.doJsonFetch` here?

shervElmi replied:
Unfortunately, we cannot use `frmDom.doJsonFetch` in this context because it is not designed to fetch data from external APIs or URLs.

Crabcyborg replied:
Thanks for looking into it! That's fine then.

Crabcyborg commented on js/form-templates/src/shared/url.js:
```diff
@@ -0,0 +1,65 @@
+ * @param {string} paramName The name of the query parameter to retrieve.
+ * @return {string|null} The value associated with the specified query parameter name, or null if not found.
+ */
+export const getQueryParam = ( paramName ) => {
```
Can this be written as `export const getQueryParam = urlParams.get`? I don't see a need for the extra arrow function.

shervElmi replied:
Thank you for your suggestion to simplify the `getQueryParam` function. While we can't directly assign `urlParams.get` due to context issues, I've refactored both `getQueryParam` and `hasQueryParam` to be more concise while still maintaining the correct context.

Crabcyborg commented on js/form-templates/src/events/saveCodeButtonListener.js:
```diff
@@ -0,0 +1,143 @@
+
+	try {
+		// Perform the POST request
+		const response = await fetch( ajaxurl, {
```
Can we use `frmDom.doJsonPost` here?

shervElmi replied:
I've reduced the scope of the `try/catch` blocks throughout the package for better readability and error handling. Regarding using `frmDom.doJsonPost`, I attempted to refactor it but encountered inconsistent results, so I've kept the current implementation for stability.

Crabcyborg commented on js/form-templates/src/events/favoriteButtonListener.js:
```diff
@@ -0,0 +1,167 @@
+			templatesList;
+
+		// Toggle twin template's favorite status
+		twinFeaturedTemplate = templateList?.querySelector(
```
Can we just exit early if there is no templateList? This is a lot harder to read with all of the nullable checks.

Crabcyborg commented on js/form-templates/src/elements/emptyStateElement.js:
```diff
@@ -0,0 +1,71 @@
+export function createEmptyStateElement() {
+	const button = a({
+		className: 'button button-primary frm-button-primary',
+		href: '#'
```
The `a` function will use `href: '#'` by default. There's no need to pass it.

Crabcyborg commented on classes/helpers/FrmFormsHelper.php:
```diff
@@ -1674,27 +1569,14 @@ public static function reserved_words() {
 	 * @param array $args
 	 * @return int
 	 */
 	public static function available_count( $templates, $args ) {
```
@shervElmi This function should still return an int. A deprecated function's return type shouldn't change. That would introduce side effects if it were being referenced somewhere.

Crabcyborg commented:
@shervElmi The new view modal is very broken when I have this branch checked out.

Can you look into that? That modal shouldn't change with this update.

<img width="582" alt="Screen Shot 2023-09-29 at 9 36 56 AM" src="https://github.com/Strategy11/formidable-forms/assets/9134515/2281286f-a47d-4d57-b9c9-e322a2aca69d">

Crabcyborg commented:
- [ ] Hitting enter on this rename pop up does not submit the form. When I press enter, it should trigger the rename button.

<img width="456" alt="Screen Shot 2023-09-29 at 10 45 01 AM" src="https://github.com/Strategy11/formidable-forms/assets/9134515/09430393-429f-4bb9-8ec7-7a64e7dfc907">

Crabcyborg commented:
- [ ] We should assign a form key on rename. Currently all forms have "noname" as a form key.

<img width="918" alt="Screen Shot 2023-09-29 at 10 45 54 AM" src="https://github.com/Strategy11/formidable-forms/assets/9134515/e75fc890-0672-47b0-834d-e373a278c908">

Crabcyborg commented:
Clicking cancel on the rename pop up triggers the form to save and reload. This feels very wrong to me.

I feel like the "Rename" button should never be disabled and just labelled "Save". Clicking cancel should just cancel the pop up.

![NEc94Tor0F](https://github.com/Strategy11/formidable-forms/assets/9134515/9c809ffc-4778-4d3f-ac11-0e25e40475f6)

Crabcyborg commented:
Switching tabs I am never prompted to name a form.

I feel like when I return to save the form, or even when saving the form from the settings tab for the first time, it would be nice to get prompted to set a form name. But once I switch tabs, I never get prompted.

![rdKWIh8v9L](https://github.com/Strategy11/formidable-forms/assets/9134515/8c244879-df7b-444b-926f-f586b1da6d3e)

garretlaxton commented:
What if a user selects a template, but they don't have the add-on installed to support it? Example, if I select an AI template but I don't have the AI add-on installed. 

I feel like we should add a pop-up that gives the option to install that add-on first.

Crabcyborg commented on js/form-templates/src/events/createFormButtonListener.js:
```diff
@@ -0,0 +1,40 @@
+const onCreateFormButtonClick = ( event ) => {
+	const { createFormButton, newTemplateForm, newTemplateNameInput, newTemplateActionInput } = getElements();
+
+	newTemplateNameInput.value = __( 'No name', 'formidable' );
```
@shervElmi Can we just leave this empty?

I don't get why we're setting "No name" here. Generally we just check if it's empty and show `(no title)`, which would be a lot more consistent.
<img width="269" alt="Screen Shot 2023-10-06 at 11 05 07 AM" src="https://github.com/Strategy11/formidable-forms/assets/9134515/b39b9682-cce1-4fe5-ab5d-1525593c3d5f">

shervElmi commented:
@Crabcyborg, thank you so much for your valuable review. I learn something new every day from your feedback.
The project is ready for another review. I've either addressed the feedback or left a comment.

tuguirazvan reviewed (changes requested):
I looked over the implementation, and the overall looks good with a few exceptions that I will detail below.

1. The description of a template is sometimes 4 lines long, which makes the rest of the cards the same height, thus leaving a lot of white space at the bottom. Here I would propose to have no more than 2 lines for the description of a template. Here I think we will have to work on reformulating the descriptions in such a way that they are on a maximum of 2 lines.
2. In an isolated case, the icon to add a template to favourites is outside the card.
3. In some areas you can see the tag for the paragraph.
4. On the "Free Templates" tab, we still have the locked icon.



![CleanShot 2023-10-12 at 11 17 22@2x](https://github.com/Strategy11/formidable-forms/assets/8666554/16f2f162-5984-49f6-906e-cc9885b65a55)

![image](https://github.com/Strategy11/formidable-forms/assets/8666554/34603fc2-f94a-4dde-91c4-dd5500c632b4)

![image](https://github.com/Strategy11/formidable-forms/assets/8666554/3bd956ae-8cb1-4757-988a-97d33ed0c0c0)

![CleanShot 2023-10-12 at 11 31 46@2x](https://github.com/Strategy11/formidable-forms/assets/8666554/42f5afd0-0b25-4ee5-af3f-5b641f8cbe34)

shervElmi commented:
@Crabcyborg, I don't think the PHPUnit test is related to my PR. Could you please check it? Thanks!

Crabcyborg commented:
@shervElmi I tried running tests locally for a few branches. It passed in all of them - master and this branch included.

It looks like the test is a bit flaky. It just failed randomly. When I re-ran it, it passed.

shervElmi commented:
Thanks for your review, @tuguirazvan.

I've addressed the issues you pointed out. You can see the fixed version in the QA URL provided in this PR's description.

Fixed version for numbers 1 and 3:

<p align="center">
  <kbd>
    <img width="1753" alt="image" src="https://github.com/Strategy11/formidable-forms/assets/69119241/342fe2e9-cfdd-4246-93c6-b095faf4edd3">
  </kbd>
</p>

Fixed version for number 2:

<p align="center">
  <kbd>
    <img width="973" alt="image" src="https://github.com/Strategy11/formidable-forms/assets/69119241/1d0f87f9-3bd0-4398-9629-2082748f5472">
  </kbd>
</p>

About number 4 (On the "Free Templates" tab, we still have the locked icon): This is correct because we show a "Leave your email address" modal to users. We then send them a code, and after that, the templates will unlock.

<p align="center">
  <kbd>
    <img width="1660" alt="image" src="https://github.com/Strategy11/formidable-forms/assets/69119241/beecd357-ab23-4696-bb2f-d38d56e2ea82">
  </kbd>
</p>

stephywells commented on images/form-templates/lock.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" width="21" height="24" fill="none" viewBox="0 0 21 24"><g fill="#4199FD" clip-path="url(#a)"><path d="M0 12a3 3 0 0 1 3-3h15a3 3 0 0 1 3 3v9a3 3 0 0 1-3 3H3a3 3 0 0 1-3-3v-9Z"/><path d="M10.5 3a3.75 3.75 0 0 0-3.75 3.75V9h-3V6.75a6.75 6.75 0 0 1 13.5 0V9h-3V6.75A3.75 3.75 0 0 0 10.5 3Z" opacity=".4"/></g><defs><clipPath id="a"><path fill="#fff" d="M0 0h21v2 ...
```
I think we'll want to added in a more reusable way. Can we move this into our icons.svg file and set the color with css?

Nice job minimizing this!!

shervElmi replied:
Resolved in: [Optimize CSS and Refine Modals](https://github.com/Strategy11/formidable-forms/pull/1304/commits/9f10a97d38e626d022aec07352aa4d977e53fb84) commit.

stephywells commented on classes/views/form-templates/modals/leave-email-modal.php:
```diff
@@ -0,0 +1,77 @@
+
+$defaults = array(
+	'api_url'     => 'https://sandbox.formidableforms.com/api/wp-json/frm/v2/forms/freetemplates?return=html&exclude_script=jquery&exclude_style=formidable-css',
+	'title'       => esc_html__( 'Get 10+ Free Form Templates', 'formidable' ),
```
Let's update this to 25+

stephywells reviewed (changes requested):
This is looking very nice and polished @shervElmi ! Very nice job on matching the designs!

It looks like you may have been misled on what the "custom" section is. Sorry about that!!

It would be great if you could go through all the css in the form-templates.css to see what we can do to keep the whole project consistent. I think there are some overrides we don't want, and some that would be good to apply plugin-wide. I left a couple comments as I started reviewing it, but I think it would be helpful if you review that in more detail since I think it would be faster since you know what everything is.

And FYI... I'd like some free templates to be unlocked by default. I'll look into updating this on our server end.

stephywells commented on classes/views/form-templates/templates.php:
```diff
@@ -0,0 +1,72 @@
+if ( 'elite' !== FrmAddonsController::license_type() && ! $expired ) {
+	FrmAppHelper::show_admin_cta(
+		array(
+			'title'       => esc_html__( 'Upgrade to get all 200+ templates', 'formidable' ),
```
Can this count be dynamic?

shervElmi replied:
Addressed in the [Dynamic Extra Template Count in Banner, Upgrade, and Renew Modals](https://github.com/Strategy11/formidable-forms/pull/1304/commits/c9469a59c3f064cf8ca99afdd4029b42c5deff13) commit.

stephywells commented on js/form-templates/src/ui/showSearchResult.js:
```diff
@@ -0,0 +1,37 @@
+import { updatePageTitle } from './';
+
+/**
+ * Updates the UI to display the search results.
```
@shervElmi how hard would it be to add the applications in like we had before? Here's a design:
https://www.figma.com/file/x8Epd7CJLLY6JIyPoCMPvc/Create-Form?node-id=1883%3A7177&mode=dev

Before:
<img width="628" alt="Screenshot 2023-10-23 at 2 21 49 PM" src="https://github.com/Strategy11/formidable-forms/assets/1116876/d7598cd5-f442-40c8-bd76-075083c40810">

shervElmi replied:
It's not hard at all. I'm familiar with all parts of the templates now and can work on them quickly.

shervElmi replied:
Addressed in the [Add Application Templates](https://github.com/Strategy11/formidable-forms/pull/1304/commits/7ce4544af52f8432630669e3aaf0b621e918b591) commit.

shervElmi commented on classes/controllers/FrmFormTemplatesController.php:
```diff
@@ -0,0 +1,790 @@
+	 */
+	private static function fetch_and_format_custom_templates() {
+		// Get all published forms.
+		$published_forms = FrmForm::get_published_forms();
```
<img width="529" alt="image" src="https://github.com/Strategy11/formidable-forms/assets/69119241/37f5acee-05e2-4e3c-bcec-3f807e0cfc4d">

stephywells replied:
I saw that. The problem is that conversation isn't correct....

shervElmi replied:
Addressed in the [Update Custom Template Functionality, Implement New Design, and Resolve Save Issues](https://github.com/Strategy11/formidable-forms/pull/1304/commits/7ee7cc229058f790205cfa7d2ee3b002ef0da43e) commit.

stephywells commented on classes/helpers/FrmAppHelper.php:
```diff
@@ -1217,6 +1228,52 @@ public static function print_admin_banner( $should_show_lite_upgrade ) {
+	 *
+	 * @return void
+	 */
+	public static function show_admin_cta( $args ) {
```
I think this would fit a bit better in the FrmTipsHelper.

shervElmi replied:
Addressed in the [Move `show_admin_cta` Method to FrmTipsHelper](https://github.com/Strategy11/formidable-forms/pull/1304/commits/c40f77e2c13b32941735e86267caeaa9a97ab4c4) commit.

stephywells commented on classes/views/welcome/show.php:
```diff
@@ -31,23 +31,16 @@
-				<p><?php esc_html_e( 'Check out our powerful pre-built templates', 'formidable' ); ?></p>
+				<a href="<?php echo esc_url( admin_url( 'admin.php?page=' . FrmFormTemplatesController::PAGE_SLUG ) ); ?>" target="_blank">
+					<!-- TODO: Get the actual text from Steph -->
+					<h3><?php esc_html_e( 'Create a New Form or Use Template', 'formidable' ); ?></h3>
```
I think "Create a New Form" is fine here.

And let's replace "our" with "a" and we're great here!

stephywells commented on classes/views/form-templates/templates.php:
```diff
@@ -0,0 +1,71 @@
+			'title'       => esc_html__( 'Get Super Powers with 200+ Pre-built Forms', 'formidable' ) . ' 🦸',
+			'description' => esc_html__( 'Unleash the potential of hundreds of additional form templates and save precious time. Upgrade today for unparalleled form-building capabilities.', 'formidable' ),
+			'link_text'   => esc_html__( 'Upgrade to PRO', 'formidable' ),
+			'link_url'    => $upgrade_link,
```
Is it possible for this link to be dynamic? It would be really nice if the utm could change so we know which templates are the best. Right now we have this:
https://formidableforms.com/lite-upgrade/?utm_source=WordPress&utm_medium=form-templates&utm_campaign=liteplugin&utm_content=upgrade

Could it be like this?
https://formidableforms.com/lite-upgrade/?utm_source=WordPress&utm_medium=form-templates&utm_campaign=liteplugin&utm_content=upgrade-advanced-mortgage-calculator-form (with the template slug)

It would be great if we could add utms on the "View Demo" link too. Like this:
https://formidableforms.com/form-templates/advanced-mortgage-calculator-form/?utm_source=WordPress&utm_medium=form-templates&utm_campaign=liteplugin&utm_content=advanced-mortgage-calculator-form

shervElmi replied:
Addressed in the [Dynamic UTM Parameters in Links for Improved Tracking](https://github.com/Strategy11/formidable-forms/pull/1304/commits/aea6becfef4d057d72e5b0911d65d12c4c8a7fe9) commit.

shervElmi commented:
Hey @stephywells,

I've addressed the following feedback with the "[Optimize CSS and Refine Modals](https://github.com/Strategy11/formidable-forms/pull/1304/commits/9f10a97d38e626d022aec07352aa4d977e53fb84)" commit:

- **Design Match:** Adjusted text color and box sizes to fit our design. Icons set to 32px, and template names to 14px.
- **Hover Design:** Implemented as shown in our Figma.
- **Design Consistency:** Used --grey-900 color for a consistent look.
- **Font Weight:** Kept h1's at a 400 font-weight for uniformity.
- **CSS Refactoring:** Expanded style usage across different areas, matched search box styling with other pages, and moved icons into `icons.svg`.
- **File Naming:** Renamed to `list.php` for easier navigation.
- **CSS Optimization:** Really worked on improving our modals and other areas. Added more classes like "frm-mb-0" and "frm-text-sm", and brought in more general classes for better flexibility.

Looking forward to your feedback, and I'm ready for any more changes if needed.

shervElmi commented on classes/views/form-templates/index.php:
```diff
@@ -0,0 +1,76 @@
+			</div>
+
+			<div id="post-body-content" class="frm_hidden">
+				<span id="frm-form-templates-create-form">
```
Resolved in: [Optimize CSS and Refine Modals](https://github.com/Strategy11/formidable-forms/pull/1304/commits/9f10a97d38e626d022aec07352aa4d977e53fb84) commit.

shervElmi commented on css/admin/form-templates.css:
```diff
@@ -0,0 +1,582 @@
+}
+
+#frm-form-templates-create-form:hover {
+	background-color: var(--grey-50);
```
Resolved in: [Optimize CSS and Refine Modals](https://github.com/Strategy11/formidable-forms/pull/1304/commits/9f10a97d38e626d022aec07352aa4d977e53fb84) commit.

shervElmi commented on css/admin/form-templates.css:
```diff
@@ -0,0 +1,582 @@
+}
+
+#frm-form-templates-empty-state .button {
+	font-weight: 600;
```
Resolved in: [Optimize CSS and Refine Modals](https://github.com/Strategy11/formidable-forms/pull/1304/commits/9f10a97d38e626d022aec07352aa4d977e53fb84) commit.

shervElmi commented on css/admin/form-templates.css:
```diff
@@ -0,0 +1,582 @@
+
+#frm-form-templates-page .frm-form-templates-title {
+	font-size: var(--text-sm);
+	color: #000;
```
Resolved in: [Optimize CSS and Refine Modals](https://github.com/Strategy11/formidable-forms/pull/1304/commits/9f10a97d38e626d022aec07352aa4d977e53fb84) commit.

shervElmi commented on css/admin/form-templates.css:
```diff
@@ -0,0 +1,582 @@
+ *
+ * This section defines the basic typography and layout styles for the Form Templates page.
+ */
+#frm-form-templates-page h1,
```
Resolved in: [Optimize CSS and Refine Modals](https://github.com/Strategy11/formidable-forms/pull/1304/commits/9f10a97d38e626d022aec07352aa4d977e53fb84) commit.

shervElmi commented on css/admin/form-templates.css:
```diff
@@ -0,0 +1,582 @@
+}
+
+/* Sidebar Scrollbar */
+#frm-form-templates-page .frm-scrollbar-wrapper {
```
Resolved in: [Optimize CSS and Refine Modals](https://github.com/Strategy11/formidable-forms/pull/1304/commits/9f10a97d38e626d022aec07352aa4d977e53fb84) commit.

shervElmi commented on css/admin/form-templates.css:
```diff
@@ -0,0 +1,582 @@
+	background: var(--grey-400); /* Show scrollbar on hover */
+}
+
+/* Search Box */
```
Resolved in: [Optimize CSS and Refine Modals](https://github.com/Strategy11/formidable-forms/pull/1304/commits/9f10a97d38e626d022aec07352aa4d977e53fb84) commit.

shervElmi commented on classes/views/form-templates/templates.php:
```diff
@@ -0,0 +1,71 @@
+<?php
```
Resolved in: [Optimize CSS and Refine Modals](https://github.com/Strategy11/formidable-forms/pull/1304/commits/9f10a97d38e626d022aec07352aa4d977e53fb84) commit.

shervElmi commented on classes/views/form-templates/templates.php:
```diff
@@ -0,0 +1,71 @@
+	die( 'You are not allowed to call this page directly.' );
+}
+?>
+<h2 id="frm-form-templates-page-title" class="frm-form-templates-title"><?php esc_html_e( 'All Templates', 'formidable' ); ?></h2>
```
Resolved in: [Optimize CSS and Refine Modals](https://github.com/Strategy11/formidable-forms/pull/1304/commits/9f10a97d38e626d022aec07352aa4d977e53fb84) commit.

shervElmi commented on classes/views/frm-forms/_publish_box.php:
```diff
@@ -10,7 +10,15 @@
 			<button class="frm_submit_form button-primary frm-button-primary frm_button_submit <?php echo esc_attr( $class ); ?>" type="button" id="frm_submit_side_top" >
-				<?php esc_html_e( 'Update', 'formidable' ); ?>
+				<?php
+				$is_new_template = FrmAppHelper::simple_get( 'new_template' );
```
Addressed in the [Update Custom Template Functionality, Implement New Design, and Resolve Save Issues](https://github.com/Strategy11/formidable-forms/pull/1304/commits/7ee7cc229058f790205cfa7d2ee3b002ef0da43e) commit.

shervElmi commented on classes/views/form-templates/modals/code-from-email-modal.php:
```diff
@@ -0,0 +1,50 @@
+}
+?>
+<div id="frm-code-from-email-modal" class="frm-form-templates-modal-item frm_hidden">
+	<div class="frm_modal_top">
```
I think it's possible to do that, but it's not a quick fix. Should I proceed with it? Also, we can add this task to my list to tackle after our release.

stephywells replied:
I made some css updates that will help. We can wait and do the rest in another update.

shervElmi commented on classes/helpers/FrmAppHelper.php:
```diff
@@ -325,6 +325,17 @@ public static function is_preview_page() {
+	 *
+	 * @return bool True if the current page is the form templates page, false otherwise.
+	 */
+	public static function is_form_templates_page() {
```
Addressed in the https://github.com/Strategy11/formidable-forms/pull/1304/commits/8c435ff08c30f57d1fda1cb53fbd2004e65ba0fb commit.

shervElmi commented on css/admin/form-templates.css:
```diff
@@ -0,0 +1,448 @@
+ * Defines the layout and styling for the sidebar, including scroll behavior.
+ */
+#frm-form-templates-page .frm-right-panel {
+	flex-basis: var(--small-sidebar);
```
addressed in the https://github.com/Strategy11/formidable-forms/pull/1304/commits/211a0ab3a0ec84687bd7bc7b40491eab753670bd commit.

shervElmi commented on css/admin/form-templates.css:
```diff
@@ -0,0 +1,448 @@
+ * Focuses on the main content area, laying out the various elements and components.
+ */
+#frm-form-templates-page #post-body-content {
+	border-left-color: var(--grey-300);
```
Addressed in the https://github.com/Strategy11/formidable-forms/pull/1304/commits/2308762bb40187ec174dfaab998c6ccc361af290 commit.

shervElmi commented on css/frm_admin.css:
```diff
@@ -1558,16 +1635,33 @@ input.frm_insert_in_template {
+}
+
+.frm-gap-xs-minus-2 {
+	gap: calc(var(--gap-xs) - 2px);
```
Addressed in the https://github.com/Strategy11/formidable-forms/pull/1304/commits/f6756248568952ca89cffd3083ca8e9f189ff180 commit.

stephywells commented on classes/views/form-templates/modals/code-from-email-modal.php:
```diff
@@ -0,0 +1,50 @@
+	die( 'You are not allowed to call this page directly.' );
+}
+?>
+<div id="frm-code-from-email-modal" class="frm-form-templates-modal-item frm_hidden">
```
I'm not able to really see much in that video. I'm not sure what you are showing.

stephywells replied:
@shervElmi I just updated two modals to show you what I mean.
https://github.com/Strategy11/formidable-forms/pull/1304/commits/f74e2b832bc1f4f25c5cc14a375a4fb6f02ce656#diff-3262ede1243c80ba8b06b9d9ec5031f4c414a4e77a19c3a2c24fe42548b8c932

Can you please update the rest? This changes the style to match what is in the rest of the plugin. We don't want to make css updates that will make these new parts look different from the rest of the plugin.

shervElmi replied:
This is the problem I was talking about. We need to use more CSS overlays for responsive devices:
![image](https://github.com/Strategy11/formidable-forms/assets/69119241/e2f3f33a-c0a8-48d9-9b6b-8574fc1a8e25)

And other parts are not as clean now, like the spacing solutions for each item, etc.:
![image](https://github.com/Strategy11/formidable-forms/assets/69119241/fb086da2-2df0-4033-b274-9c6ef3714b50)

shervElmi replied:
Addressed in commit [c88c430](https://github.com/Strategy11/formidable-forms/pull/1304/commits/c88c4306eeb65e09ef25fd32c4ecb70dd91a4d26).

After our changes, we need to adjust the modals again. You can see them in the following images:

![Modal Adjustment 1](https://github.com/Strategy11/formidable-forms/assets/69119241/f85da339-d764-46b1-9b5d-320e74bfa00d)
![Modal Adjustment 2](https://github.com/Strategy11/formidable-forms/assets/69119241/ea490c62-e7f9-4df1-acc1-b2c474959f32)
![Modal Adjustment 3](https://github.com/Strategy11/formidable-forms/assets/69119241/59d53b0a-81b4-4f02-81e9-103ec5265b89)

shervElmi replied:
I have already applied them correctly, and our attention is needed on other fixes. Here are the related issues I've resolved in the following commits:

1. Fix: Hide upgrade modal footer in non-upgrade modals - [Commit Link](https://github.com/Strategy11/formidable-forms/pull/1304/commits/322004823c36a647e48c508c5e6a3fde8ebabccd)
2. Fix: Retain email modal styles, update title to 20+ - [Commit Link](https://github.com/Strategy11/formidable-forms/pull/1304/commits/0c58eb1a12912a8f86aa989e48a1cf56d68f1055)
3. Fix: Update subscribe form response message on empty inbox page - [Commit Link](https://github.com/Strategy11/formidable-forms/pull/1304/commits/6261a6d4aa562603ea1fb3614ab77ae1062bce6a)
   - **Before:**
    <kbd><img width="539" alt="image" src="https://github.com/Strategy11/formidable-forms/assets/69119241/23f7a193-3543-479e-958c-834710794be1"></kbd>
    <kbd><img width="526" alt="image" src="https://github.com/Strategy11/formidable-forms/assets/69119241/33b714b1-5cd0-477e-8118-0b4c554b6dd5"></kbd>

   - **After:**
    <kbd><img width="520" alt="image" src="https://github.com/Strategy11/formidable-forms/assets/69119241/bb7eaea8-53cc-449d-bf5c-b72d4fd508fe"></kbd>
    <kbd><img width="519" alt="image" src="https://github.com/Strategy11/formidable-forms/assets/69119241/b946c21f-5c89-4c92-bce6-704fc0b28450"></kbd>

stephywells replied:
Thanks so much!

stephywells commented on classes/views/frm-forms/_publish_box.php:
```diff
@@ -10,7 +10,15 @@
 			<button class="frm_submit_form button-primary frm-button-primary frm_button_submit <?php echo esc_attr( $class ); ?>" type="button" id="frm_submit_side_top" >
-				<?php esc_html_e( 'Update', 'formidable' ); ?>
+				<?php
+				$is_new_template = FrmAppHelper::simple_get( 'new_template' );
```
A helper file is an ideal place for functions called from inside a view file. I'd add it in FrmFormsHelper.

shervElmi replied:
Thank you for your guidance, Steph! I've addressed this in the commit at https://github.com/Strategy11/formidable-forms/pull/1304/commits/8fe8f738e17c75f2f7901abc62a3e581c665ac9e.

stephywells commented on css/frm_admin.css:
```diff
@@ -1558,16 +1635,33 @@ input.frm_insert_in_template {
 	gap: var(--gap-xs);
 }
 
+.frm-gap-xs-2xs {
```
Since the padding for left and right is different from top and bottom, it should be two separate classes. One for x and one for y.

stephywells replied:
You might be asking about a different commit. For any gap sizes that aren't in our list, we'll want to choose the closest option that looks the best. We want to avoid adding padding sizes that are not included in our options.

shervElmi replied:
Addressed in this commit: [c88c430](https://github.com/Strategy11/formidable-forms/pull/1304/commits/c88c4306eeb65e09ef25fd32c4ecb70dd91a4d26). The spaces between buttons are now as shown below. Does this look fine to you?

<img width="708" alt="image" src="https://github.com/Strategy11/formidable-forms/assets/69119241/5da8ad97-906a-4f2f-a0e8-a1ece9d7b925">

shervElmi commented on js/form-templates/src/utils/animationEffects.js:
```diff
@@ -8,10 +8,9 @@ import { isVisible, show } from './';
  * @return {void}
  */
-export const fadeIn = ( element, fadingClass = `${PREFIX}-flex` ) => {
+export const fadeIn = ( element ) => {
```
Steph, please don't remove the fadingClass parameter. It's very useful for controlling the layout during an animation. For instance, when we need to fade in a grid layout, we need to add a class that has `display: grid` property. This could also apply to other display properties like `inline-block`, `block`, `flex`, etc.

stephywells replied:
@shervElmi we should be able to control this with css, right?

stephywells replied:
It looks like the problem was this css:
```
#frm-form-templates-page .frm_hidden {
	display: none !important;
}
```

Once I removed that and made css less greedy, I think it prevents the need for this workaround.

stephywells replied:
I'm not seeing any difference in the end result. Can you record a quick screen capture to show the issue you are seeing with this change?

shervElmi replied:
In our case, it doesn't make a difference, but about two months ago, I fixed a UI issue related to this problem. It's not a big deal, and we can go ahead and remove that.

stephywells replied:
Thanks for double checking!

stephywells commented on css/admin/form-templates.css:
```diff
@@ -0,0 +1,448 @@
+	border-left-color: var(--grey-300);
+}
+
+#frm-form-templates-page .frm-form-templates-grid-layout {
```
I just updated this in https://github.com/Strategy11/formidable-forms/pull/1304/commits/f74e2b832bc1f4f25c5cc14a375a4fb6f02ce656#diff-1554f4f6ac22ca08073445dd104449401dfabe09b4842ad1834ffc1150d3efc5

shervElmi replied:
Cool, as I mentioned, I was concerned about further CSS overwrites! Thanks for your updates ☺️

shervElmi replied:
This is the problem I was talking about. We need to use more CSS overlays for responsive devices:
![image](https://github.com/Strategy11/formidable-forms/assets/69119241/e2f3f33a-c0a8-48d9-9b6b-8574fc1a8e25)

And other parts are not as clean now, like the spacing solutions for each item, etc.:
![image](https://github.com/Strategy11/formidable-forms/assets/69119241/fb086da2-2df0-4033-b274-9c6ef3714b50)

stephywells replied:
Does it help if you add this?
```
@media only screen and (min-width: 2550px) {
	#frm-form-templates-page .frm_grid_container .frm4 {
		grid-column: span 3;
	}
}
```

shervElmi replied:
Yes, thanks!

stephywells replied:
Actually... nevermind. I think I'm trying to force this one too much. I'll revert the grid class change.

stephywells replied:
This grid change is reverted. Thanks for sticking to your guns!

shervElmi replied:
Thank you so much! I'm very happy now! ❤️

stephywells commented:
> Do I only need to show the title here, or should I also return the image?

@shervElmi I'm not sure if this was related to my changes or not, but I removed the js that was hiding the title.
https://github.com/Strategy11/formidable-forms/pull/1304/commits/b18ea7405b34ae76c71e04ad25e2764b7a90326a#diff-23fb66f34c625a76ac7564df177d3ab58c86b1f5d071f745aa50e028fb0209c4L9382

shervElmi commented on classes/views/form-templates/modals/create-template-modal.php:
```diff
@@ -0,0 +1,71 @@
+<div id="frm-create-template-modal" class="frm-form-templates-modal-item frm_hidden">
+	<div class="frm_modal_top">
+		<div class="frm-modal-title">
+			<h2 class="frm-font-semibold frm-m-0"><?php esc_html_e( 'Create New Template', 'formidable' ); ?></h2>
```
Great update! 🚀

shervElmi commented on classes/views/form-templates/list.php:
```diff
@@ -68,8 +68,6 @@
 </ul>
 
 <div id="frm-form-templates-custom-list-section" class="frm-flex-col frm-gap-sm">
-	<h3 id="frm-form-templates-custom-list-title" class="frm-form-templates-title frm-text-xs frm-font-medium frm-mt-xs frm-mb-0 frm_hidden"><?php esc_html_e( 'Custom List', 'formidable' ); ?></h3>
```
I used this title when we had both custom and regular favorites. This title was shown because regular favorites and custom favorites are separate lists. Look at the before and after code:

### Before:
![Before Changes](https://github.com/Strategy11/formidable-forms/assets/69119241/46a17309-3ba5-4a11-adfd-84409dfbae81)

### After:
![After Changes](https://github.com/Strategy11/formidable-forms/assets/69119241/db66c306-069e-47ea-b762-618ea89af147)

shervElmi replied:
Thank you, Steph. It's fixed now! 🚀

Crabcyborg commented on classes/controllers/FrmFormTemplatesController.php:
```diff
@@ -0,0 +1,864 @@
+		// Fetch and format custom templates.
+		self::fetch_and_format_custom_templates();
+
+		// Retrieve and set the templates.
```
Some of these comments aren't really necessary. When something is as straight forward as copying the function name, I think it's fine to leave some comments out.

shervElmi replied:
Addressed in the https://github.com/Strategy11/formidable-forms/pull/1304/commits/3ee59a204c9a4c25fdfc4b5704fb9c321f3873a5 commit.

Crabcyborg commented on classes/controllers/FrmHooksController.php:
```diff
@@ -188,6 +188,14 @@ public static function load_admin_hooks() {
+		add_action( 'admin_footer', 'FrmFormTemplatesController::render_modal' );
+		add_filter( 'frm_form_nav_list', 'FrmFormTemplatesController::append_new_template_to_nav', 10, 2 );
+		add_action( 'admin_enqueue_scripts', 'FrmFormTemplatesController::enqueue_assets', 15 );
+		add_action( 'admin_enqueue_scripts', 'FrmFormTemplatesController::dequeue_scripts', 15 );
```
Can we add some of these actions later, when we already know it's the form templates page, rather than always adding these actions and checking and returning inside of those functions?

It helps get rid of the overhear / memory load if we don't add as many actions when we don't need them. And for form templates, these are needed very seldom, only on a specific page.

shervElmi replied:
Addressed in the https://github.com/Strategy11/formidable-forms/pull/1304/commits/d7074eba4304090cb289a0c77b0f5f77a690d7f3 commit.

Crabcyborg commented on classes/helpers/FrmFormsHelper.php:
```diff
@@ -1768,4 +1737,42 @@ public static function maybe_add_sanitize_url_attr( $url, $form_id ) {
+	 * @param array $args
+	 * @return int
+	 */
+	public static function available_count( $templates, $args ) {
```
We're probably fine just returning `0` and dropping the body of this.

shervElmi replied:
Addressed in the https://github.com/Strategy11/formidable-forms/pull/1304/commits/c3333167181d60b7eaaff7fd959057ec76a91e65 commit.

Crabcyborg commented on classes/views/form-templates/modals/create-template-modal.php:
```diff
@@ -0,0 +1,68 @@
+					$form_description = FrmAppHelper::kses( $form->description, array( 'a', 'i', 'span', 'use', 'svg' ) );
+					?>
+					<option value="<?php echo esc_attr( $form->id ); ?>" data-name="<?php echo esc_attr( $form->name ); ?>" data-description="<?php echo esc_attr( $form_description ); ?>">
+						<?php echo esc_html( empty( $form->name ) ? __( '(no title)', 'formidable' ) : $form->name ); ?>
```
Maybe instead of adding another `(no title)` we use `FrmFormsHelper::edit_form_link_label( $form )` here? It adds some truncation as well, but anything past 40 characters should probably be truncated anyway so that's a plus.

Crabcyborg commented on js/src/form-templates/ui/showSelectedCategory.js:
```diff
@@ -0,0 +1,183 @@
+		elementsToShow.push( customTemplatesSection );
+		elementsToShow.push( customTemplatesList );
+
+		0 === favoritesCount.default ?
```
Inline ternary conditions like this are not how we structure our `if`/`else` conditions. Can we just do an if/else here? It's easier to read.

shervElmi replied:
Addressed in the https://github.com/Strategy11/formidable-forms/pull/1304/commits/209fe859ca38f699c8f18b363b6576e9549ac458 commit.

Crabcyborg commented on js/src/form-templates/templates/applicationTemplates.js:
```diff
@@ -0,0 +1,41 @@
+	}
+
+	doJsonFetch( 'get_applications_data&view=templates' ).then( ( data ) => {
+		setupApplicationTemplates( data );
```
This looks like it can be simplified to just:
```
doJsonFetch( 'get_applications_data&view=templates' ).then( setupApplicationTemplates );
```

Crabcyborg commented on js/src/common/utilities/requestQueue.js:
```diff
@@ -0,0 +1,10 @@
+ * @param {function(): Promise<any>} task A function that returns a promise.
+ * @return {Promise<any>} The new last promise in the queue.
+ */
+export const addToRequestQueue = task => lastPromise = lastPromise.then( () => task() ).catch( () => task() );
```
This looks like it would be a bit more flexible and smaller if it was just written as:

```
export const addToRequestQueue = task => lastPromise = lastPromise.then( task ).catch( task );
```

That way if you ended up wanting to pass a parameter as well, you could, and there are fewer functions involved.

shervElmi replied:
Addressed in the https://github.com/Strategy11/formidable-forms/pull/1304/commits/da3b84f13351f448efdfc9d86073c7cf7c6c261e commit.

Crabcyborg commented on classes/views/form-templates/template.php:
```diff
@@ -0,0 +1,142 @@
+				if ( $template['description'] ) {
+					echo FrmAppHelper::kses( $template['description'], array( 'a', 'i', 'span', 'use', 'svg' ) ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+				} else {
+					echo '<i>' . esc_html__( 'No description', 'formidable' ) . '</i>';
```
Do we really want to show "No description"? Does it make more sense to just omit this HTML?

stephywells replied:
I was debating on this too. I removed it at one point, then decided it feels too empty without something there. I think we can leave it. I've got a commit incoming to replace that with the date the template was created. Might as well put something there that's a little helpful.

Crabcyborg commented on classes/views/form-templates/template.php:
```diff
@@ -0,0 +1,142 @@
+	$class_names[] = 'frm-form-templates-favorite-item';
+}
+
+if ( $is_custom_template ) {
```
There are a lot of if statements in here for a view file. Can we move some of this into a help function? One that got all of these `$class_names` seems like it would be useful.

Crabcyborg reviewed (changes requested):
@shervElmi None of the UI gets focus on tab.

This page is totally unusable in a screen reader.

Can we make sure that customers are able to add templates and create forms with a screen reader and using only the keyboard?

shervElmi commented:
> @shervElmi None of the UI gets focus on tab.
> 
> This page is totally unusable in a screen reader.
> 
> Can we make sure that customers are able to add templates and create forms with a screen reader and using only the keyboard?

That's a great point! Thank you for bringing that up. I'll work on it today!

stephywells commented:
> It acts as a preloader for the Form Template page, which is significant, and I agree we should remove it as you suggested.
> 
> However, please take a look at the sidebar without the fadeIn effect, particularly the "Free Templates" item and the updating of the "Available Templates count number" via JavaScript. What are your thoughts on this?

@shervElmi we can still use the fade in effect. We just need to add a new class like this to use in place of frm_hidden in some cases:
```
.js .frm-hide-js {
display: none;
}
```

shervElmi commented:
> > It acts as a preloader for the Form Template page, which is significant, and I agree we should remove it as you suggested.
> > However, please take a look at the sidebar without the fadeIn effect, particularly the "Free Templates" item and the updating of the "Available Templates count number" via JavaScript. What are your thoughts on this?
> 
> @shervElmi we can still use the fade in effect. We just need to add a new class like this to use in place of frm_hidden in some cases:
> 
> ```
> .js .frm-hide-js {
> display: none;
> }
> ```

Thank you for your input on this matter. I've addressed it in the commit https://github.com/Strategy11/formidable-forms/pull/1304/commits/0e3c413d11a3973d448c02566eb6d8485ae27891.

shervElmi commented on classes/controllers/FrmAppController.php:
```diff
@@ -1118,10 +1072,11 @@ public static function add_admin_footer_links() {
+		$is_form_templates = FrmFormTemplatesController::is_templates_page();
 		$is_not_formidable_branding = ! FrmAppHelper::is_formidable_branding();
 
 		// Exit if any of the above conditions are met
```
Addressed in the https://github.com/Strategy11/formidable-forms/pull/1304/commits/aa7a09cfae85879d4660d46f1d87cd7b6801234a commit.

stephywells replied:
I simplified this a bit in https://github.com/Strategy11/formidable-forms/pull/1304/commits/0961031bd0ce116d2127dcac225b2738fc710668 so we don't have to keep adding conditions.

shervElmi commented:
Thank you for your contributions, Steph. I've noticed an issue with the banner style! it appears to be incorrect now. This seems to have happened after the recent changes you made to the cta method in your last commits.

<img width="1502" alt="image" src="https://github.com/Strategy11/formidable-forms/assets/69119241/1f091a52-4e46-4aa8-9140-f5a84b740a9f">

shervElmi commented:
@stephywells, I've implemented a small change to fix the CTA banner's view style, following our recent work on it. You can view this modification in the following commit: https://github.com/Strategy11/formidable-forms/pull/1304/commits/bdb35fd9978650c80ddd3ad2c7efae851acc1222

Crabcyborg commented on js/formidable_blocks.js:
```diff
@@ -1 +1,1923 @@
-(()=>{var e={877:(e,t)=>{"use strict";Object.defineProperty(t,"__esModule",{value:!0});var r=function(){function e(e,t){for(var r=0;r<t.length;r++){var n=t[r];n.enumerable=n.enumerable||!1,n.configurable=!0,"value"in n&&(n.writable=!0),Object.defineProperty(e,n.key,n)}}return function(t,r,n){return r&&e(t.prototype,r),n&&e(t,n),t}}(),n=wp.element.Component,o=wp.components.Dashicon,l=function(e){f ...
```
@shervElmi Why is your update changing these other build files?

It looks like the new output isn't minified either?

Crabcyborg replied:
Can we make sure that everything is at least minified?

stephywells replied:
@shervElmi it doesn't look like this will work. We need to update the config so they minify automatically.

shervElmi replied:
Thanks for your input, Steph! Could you let me know where the config is located? Please note that I've already added Form Templates to the Webpack configuration.

stephywells replied:
Looks like "npm run build" is handling the minification fine. I added a commit to get rid of an extra form-templates.js.license.txt file that was getting created.

I also updated dependencies to get rid of critical issues.

shervElmi replied:
Thanks so much, Steph!

stephywells commented:
> I've implemented a small change to fix the CTA banner's view style, following our recent work on it. You can view this modification in the following commit: [bdb35fd](https://github.com/Strategy11/formidable-forms/commit/bdb35fd9978650c80ddd3ad2c7efae851acc1222)

Thank you for fixing it!

stephywells reviewed (approved):
I think there is still more we will want to do to make some of this js reusable. I haven't reviewed every js file in detail, but I think the js is easier to make changes in the future since it won't be distributed.

Thanks for the great work on this @shervElmi ! I think it looks ready to go!!

shervElmi commented:
Thanks for all your work and the review on this, @stephywells! You're as great as always!

shervElmi commented on css/frm_admin.css:
```diff
@@ -2488,80 +2669,21 @@ body.frm-body-with-open-modal {
-.control-section:hover .accordion-section-title,
-#frm_new_form_modal .accordion-section.open .accordion-section-title
-{
+.control-section:hover .accordion-section-title {
```
Yes, I remember that we need them in the "Select View Type" modal on the Views page.

Crabcyborg commented:
@shervElmi I see this JS error on load.

<img width="951" alt="Screen Shot 2023-12-12 at 10 29 26 AM" src="https://github.com/Strategy11/formidable-forms/assets/9134515/7c81e5b9-22a9-44ac-a8ae-e0541a9d4380">

Crabcyborg commented:
@shervElmi I found a styling issue with the entries search page introduced with this branch.

**Before**
<img width="506" alt="Screen Shot 2023-12-12 at 10 38 23 AM" src="https://github.com/Strategy11/formidable-forms/assets/9134515/7d4943bc-2fbf-4e66-af33-506dc4a982df">

**After**
<img width="467" alt="Screen Shot 2023-12-12 at 10 38 28 AM" src="https://github.com/Strategy11/formidable-forms/assets/9134515/2d3d005e-0aa6-4629-88a1-f25cdfe71252">

shervElmi commented:
> @shervElmi I see this JS error on load.
> 
> <img alt="Screen Shot 2023-12-12 at 10 29 26 AM" width="951" src="https://private-user-images.githubusercontent.com/9134515/289893775-7c81e5b9-22a9-44ac-a8ae-e0541a9d4380.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIzOTI4NzQsIm5iZiI6MTcwMjM5MjU3NCwicGF0aCI6Ii85MTM0NTE1LzI4OTg5Mzc3NS03YzgxZTViOS0yMmE5LTQ0YWMtYThhZS1lMDU0MWE5ZDQzODAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTJUMTQ0OTM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDk2NWQ4MTkxOTllMGQ4MjljMDZhZjRmOWMwYTVjNmNiNDJiNzVhYjUyN2ViOWE5Y2E2MmRmMWY0Yzc2ZmNjMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.B2LdTYRCB_khpOhPR8gjrri65yo-PTQ-tENHHa-O694">

Mike, I can't see this error. Could you please clear your cache and check again? If it's still there, could you give me more details about how to reproduce it?

shervElmi commented:
> I found a styling issue with the entries search page introduced with this branch.

Corrected text:

Thanks for pointing that out, Mike! I've addressed this issue in the commit https://github.com/Strategy11/formidable-forms/pull/1304/commits/521ede92675f2b6a9e0fe0e054aafb6cafe87bb7. I've also reviewed all inputs to ensure they are correct.

Crabcyborg reviewed (approved):
Thanks for updating that @shervElmi!

I think this is good to go now.

I'll try to get this merged.
